From 76236440d1f84bb85a5b60d5634499742d866c0f Mon Sep 17 00:00:00 2001
From: Konrad Leszczynski <konrad.leszczynski@intel.com>
Date: Fri, 20 Nov 2015 15:41:25 +0000
Subject: [PATCH] usb: xhci: disable U1 and U2 states for SSIC port

This patch disables the U1 and U2 states for the SSIC
port, on which the Intel modems (726x/736x) are enumerating.
It was reported that the U1/U2 modem states are not
functioning properly (wakeup problems etc.) and decided
to disable them.

Change-Id: I4cf7181b1428f2eaa9fa4bdcae2f42329dbb49af
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9670
Signed-off-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
---
 drivers/usb/host/xhci-hub.c       | 13 +++++++++++--
 drivers/usb/host/xhci-intel-cap.c |  7 +++++++
 drivers/usb/host/xhci-intel-cap.h |  1 +
 drivers/usb/host/xhci.h           |  2 ++
 4 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/host/xhci-hub.c b/drivers/usb/host/xhci-hub.c
index bc49f7a..4514644 100644
--- a/drivers/usb/host/xhci-hub.c
+++ b/drivers/usb/host/xhci-hub.c
@@ -25,6 +25,7 @@
 
 #include "xhci.h"
 #include "xhci-trace.h"
+#include "xhci-intel-cap.h"
 
 #define	PORT_WAKE_BITS	(PORT_WKOC_E | PORT_WKDISC_E | PORT_WKCONN_E)
 #define	PORT_RWC_BITS	(PORT_CSC | PORT_PEC | PORT_WRC | PORT_OCC | \
@@ -953,7 +954,11 @@ int xhci_hub_control(struct usb_hcd *hcd, u16 typeReq, u16 wValue,
 				goto error;
 			temp = readl(port_array[wIndex] + PORTPMSC);
 			temp &= ~PORT_U1_TIMEOUT_MASK;
-			temp |= PORT_U1_TIMEOUT(timeout);
+			/* Disable U1 for Intel Modem */
+			if (xhci_intel_ssic_port_check(xhci, wIndex + 1))
+				temp |= PORT_U1_TIMEOUT(PORT_U1_DISABLE);
+			else
+				temp |= PORT_U1_TIMEOUT(timeout);
 			writel(temp, port_array[wIndex] + PORTPMSC);
 			break;
 		case USB_PORT_FEAT_U2_TIMEOUT:
@@ -961,7 +966,11 @@ int xhci_hub_control(struct usb_hcd *hcd, u16 typeReq, u16 wValue,
 				goto error;
 			temp = readl(port_array[wIndex] + PORTPMSC);
 			temp &= ~PORT_U2_TIMEOUT_MASK;
-			temp |= PORT_U2_TIMEOUT(timeout);
+			/* Disable U2 for Intel Modem */
+			if (xhci_intel_ssic_port_check(xhci, wIndex + 1))
+				temp |= PORT_U2_TIMEOUT(PORT_U2_DISABLE);
+			else
+				temp |= PORT_U2_TIMEOUT(timeout);
 			writel(temp, port_array[wIndex] + PORTPMSC);
 			break;
 		case USB_PORT_FEAT_TEST:
diff --git a/drivers/usb/host/xhci-intel-cap.c b/drivers/usb/host/xhci-intel-cap.c
index 5daabff..cada8c66 100644
--- a/drivers/usb/host/xhci-intel-cap.c
+++ b/drivers/usb/host/xhci-intel-cap.c
@@ -47,6 +47,13 @@ void xhci_change_ssic_regs(struct xhci_hcd *xhci, bool enable)
 			readl(hcd->regs + SSIC_CONFIG2));
 }
 
+bool xhci_intel_ssic_port_check(struct xhci_hcd *xhci, int port)
+{
+	/* Check if device present and the port matches */
+	return (xhci->ssic_device_present) &&
+			(port == xhci->ssic_port_number);
+}
+
 int xhci_intel_need_disable_stall(struct xhci_hcd *xhci)
 {
 	struct acpi_device *acpi_dev;
diff --git a/drivers/usb/host/xhci-intel-cap.h b/drivers/usb/host/xhci-intel-cap.h
index 2733bcb..d7becb7 100644
--- a/drivers/usb/host/xhci-intel-cap.h
+++ b/drivers/usb/host/xhci-intel-cap.h
@@ -71,4 +71,5 @@ extern int xhci_intel_phy_mux_switch(struct xhci_hcd *xhci, int is_device_on);
 extern void xhci_intel_clr_internal_pme_flag(struct xhci_hcd *xhci);
 extern void xhci_intel_ssic_port_unused(struct xhci_hcd *xhci, bool unused);
 extern void xhci_change_ssic_regs(struct xhci_hcd *xhci, bool enable);
+extern bool xhci_intel_ssic_port_check(struct xhci_hcd *xhci, int port);
 extern int xhci_intel_need_disable_stall(struct xhci_hcd *xhci);
diff --git a/drivers/usb/host/xhci.h b/drivers/usb/host/xhci.h
index 42b603b..2d1c1fc 100644
--- a/drivers/usb/host/xhci.h
+++ b/drivers/usb/host/xhci.h
@@ -375,9 +375,11 @@ struct xhci_op_regs {
  */
 #define PORT_U1_TIMEOUT(p)	((p) & 0xff)
 #define PORT_U1_TIMEOUT_MASK	0xff
+#define PORT_U1_DISABLE	0xff
 /* Inactivity timer value for transitions into U2 */
 #define PORT_U2_TIMEOUT(p)	(((p) & 0xff) << 8)
 #define PORT_U2_TIMEOUT_MASK	(0xff << 8)
+#define PORT_U2_DISABLE	0xff
 /* Bits 24:31 for port testing */
 
 /* USB2 Protocol PORTSPMSC */
-- 
1.9.1

